doc/Data Plane Code Generation/ServiceDrivenEvolution.md

```diff
@@ -0,0 +1,179 @@
+# Service driven evolution
+
+As a service evolves with non-breaking changes in the REST API, it may or may not cause breaking changes in the generated SDK. For instance, adding a new method in an existing path or in a new path won't cause breaking changes in the new SDK. See details [here](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#api-versioning). 
+
+However, even no breaking changes exist from the REST API perspective, some changes like adding an optional query parameter could still cause breaking changes in the generated SDK. In order to make the SDK backward compatible, we can add customized code such as overload methods for now. We will make it automatically in the near future. The following sections will show some examples of such scenarios and how to handle them.
+
+## A method gets a new optional parameter
+
+Generated code in a V1 client with a required query parameter and an optional query parameter:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam = null, RequestContext context = null)
+{
+    ...
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    ...
+}
+```
+
+Generated code in a V2 client with a new optional query parameter:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam = null, string newParameter = null, RequestContext context = null)
+{
+    ...
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, newParameter, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    ...
+}
+```
+
+Since `context` is always the last parameter, V2 inserts the new optional parameter before `context` parameter and this could break code that works with V1 such as `client.PutRequiredOptionalAsync("requiredParam", "optionalParam", ErrorOptions.NoThrow)` (`ErrorOptions.NoThrow` was passed in as `context` parameter but now it's treated as `newParameter`). We can manually add an overload method in V2 which transforms the optional parameters in V1 to required parameters to make V2 backward compatible with V1:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam, RequestContext context)
+{
+    ...
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, null, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    ...
+}
+```
+
+**Notice:**
+Now the helper method has one more parameter, so we should call it with `newParameter` `null` as `CreatePutRequiredOptionalRequest(requiredParam, optionalParam, null, context);`.
```

Comment:
We're calling these "helper method" and "manual method" -- there's some ambiguity there.  Is there a more clear name we could use?